### ISSUE TITLE
Creating offset file takes a very long time

### ISSUE THREAD

**Issue by synhershko** [elastic/rally#643]
We have a jsonl file with ~1.5 lines, each line has a lot of characters (big JSONs). Running esrally starts with creating the offset file but that step takes a very very long time. During that time, we observe a very low CPU usage by esrally in a single core (usually 1-2%).

This happens on AWS and a SSD drive with provisioned IOPS.

For now we create a snapshot right after the offset file has been written, but why is this needed and why is it taking so long?

**Comment by ebadyano** [comment]
Hi @synhershko, Thank you for your interest in Rally! We reserve Github for bug reports and feature requests only. Please ask questions like these in the [Rally forum](https://discuss.elastic.co/c/elasticsearch/rally) instead. Thank you!

**Comment by synhershko** [comment]
Hi @ebadyano , as a long time contributor to Elastic I'm well aware of that policy and I do believe this falls under the category "bug". Can you please reopen, and let me know if I can provide any further details to let you handle this better?

cc @danielmitterdorfer

**Comment by danielmitterdorfer** [comment]
Thanks for the feedback but I don't see how this is a bug. You ask:

> For now we create a snapshot right after the offset file has been written, but why is this needed and why is it taking so long?

Rally uses the offset file as a persistent mapping table from line numbers in the document file to bytes offsets. We need this because if you bulk-index with multiple clients, each client will read one section of the document file and send documents from that section to the bulk API. Suppose you have a file with 1000 documents and want to bulk index with 4 clients:

* client 0: reads lines 1 - 250
* client 1: reads lines 251 - 500
* client 2: reads lines 501 - 750
* client 3: reads lines 751 - 1000

So apart from client 0, all other clients need to skip lines in the file. In the intial implementation we did this by reading the file from the beginning but noticed that skipping ahead can take a long time even on an SSD (see https://github.com/elastic/rally/issues/158). This means that contrary to user expectation we start to apply load gradually but instead all clients should start at the same time.

The file I/O commands only allow to skip a number of bytes, not a number of lines so we need a means means to quickly jump to the correct line. And this is what the offset file is for. It needs to be created the first time by linearly reading through the file, identifiying line breaks and creating that offset table. After this is done, the offset file is reused (unless you change the corresponding document file, then Rally will notice that and recreate the offset file).

Now we could get rid of the offset file if we have one document per client but that does not help either if you want to quickly change the number of clients so this is the design tradeoff that we made.

I hope that makes more sense now and if there are follow-up discussions we should really take them to the forums where this is more visible instead of a closed Github issue.
